airflow/models.py

```diff
@@ -2010,6 +2129,105 @@ def get(cls, key, session, deserialize_json=False):
         return v
 
 
+class XCom(Base):
+    """
+    Base class for XCom objects.
+    """
+    __tablename__ = "xcom"
+
+    id = Column(Integer, primary_key=True)
+    key = Column(String)
+    value = Column(PickleType(pickler=dill))
+    timestamp = Column(DateTime, server_default=func.current_timestamp())
+    visible_on = Column(DateTime, nullable=False)
+
+    # source information
+    task = Column(String, nullable=False)
+    dag = Column(String, nullable=False)
+
+    def __repr__(self):
+        return '<XCom "{key}" ({task} @ {visible_on})>'.format(
+            key=self.key,
+            task=self.task,
+            visible_on=self.visible_on)
+
+    @classmethod
+    @provide_session
+    def set(
+            cls,
+            key,
+            value,
+            visible_on,
+            task,
+            dag,
+            session=None):
+        """
+        Store an XCom value.
+        """
+
+        if not isinstance(task, basestring) or not isinstance(dag, basestring):
+            raise TypeError('Expected string; received {}'.format(type(task)))
+
+        session.expunge_all()
+        session.add(XCom(
+            key=key,
+            value=value,
+            visible_on=visible_on,
+            task=task,
+            dag=dag))
+        session.commit()
+
+    @classmethod
+    @provide_session
+    def get(
+            cls,
+            visible_on,
+            key=None,
+            tasks=None,
+            dags=None,
+            include_prior_dates=False,
+            limit=None,
+            session=None):
+        """
+        Retrieve an XCom value, optionally meeting certain criteria
+        """
+        filters = []
+        if key:
+            filters.append(cls.key == key)
+        if tasks:
+            filters.append(cls.task.in_(as_tuple(tasks)))
+        if dags:
+            filters.append(cls.dag.in_(as_tuple(dags)))
+        if include_prior_dates:
+            filters.append(cls.visible_on <= visible_on)
+        else:
+            filters.append(cls.visible_on == visible_on)
+
+        query = (
+            session.query(cls)
+            .filter(and_(*filters))
+            .order_by(cls.visible_on.desc(), cls.timestamp.desc())
+            .limit(limit or 1))
+
+        result = pd.read_sql(
+            sql=query.statement,
+            con=query.session.bind,
+            index_col='id',
+            parse_dates=['timestamp', 'visible_on'])
+
+        result.drop_duplicates(
```

Comment:
SO this actually turns out to be tricky. I'm not sure an XCom is actually guaranteed to be unique because users could rerun a DAG for a given `execution_date` and if we don't delete the previously-generated XComs, it would violate the constraint when the new XComs are added.

I suppose I could watch for constraint violations and then UPDATE that row rather than INSERT the new one.

Further complicating this, I couldn't find an efficient way to remove the duplicates in database-agnostic SQL (in Postgres it's simply `DISTINCT ON()`) so that's why I resorted to doing it in-memory. 

Open to any suggestions!